modules/tinymce/src/plugins/accordion/main/ts/core/Actions.ts

```diff
@@ -0,0 +1,52 @@
+import Editor from 'tinymce/core/api/Editor';
+
+import { fireInsertAccordionEvent } from '../api/Events';
+
+const nonRemovableContainers = [ 'BODY', 'TD', 'TH', 'LI', 'DIV', 'DT', 'DD' ];
+const isNonRemovableContainer = (node: Node): boolean =>
+  nonRemovableContainers.includes(node.nodeName);
+
+const validAncestorContainers =
+  nonRemovableContainers.concat([ 'P', 'H1', 'H2', 'H3', 'H4', 'H5', 'H6', 'UL', 'OL', 'TABLE', 'FIGURE', 'DETAILS', 'DL' ]);
+const isValidAncestorContainer = (node: Node): node is HTMLElement =>
+  validAncestorContainers.includes(node.nodeName);
+
+const removeBr = (node: Node): void => {
+  if (node.firstChild?.nodeName === 'BR') {
+    node.firstChild.remove();
+  }
+};
+
+const insertAccordion = (editor: Editor): void => {
+  const dom = editor.dom;
+
+  const rng = editor.selection.getRng();
+  const summaryText = rng.toString() || editor.translate('Accordion summary...');
+  const bodyText = editor.translate('Accordion body...');
+
+  const target = dom.getParent(rng.commonAncestorContainer, isValidAncestorContainer);
+
+  if (!target) {
+    return;
+  }
+
+  const details = dom.create('details', { class: 'mce-accordion', open: 'open' });
```

Comment:
I wonder if this whole insert thing could be done with the insertContent API instead since that is schema aware. You can get the inserted summary location by doing a temporary data attribute.

```
const uid = Id.generate('acc');
const summaryText = editor.dom.encode(editor.selection.getRng().toString() || editor.translate('Accordion summary...'));
const bodyText = editor.dom.encode(editor.translate('Accordion body...'));
editor.insertContent(`<details mce-accordion open><summary class="mce-accordion-summary" data-mce-id="${uid}">${summaryText}<div class="mce-accordion-body"><p>${bodyText}</p></div></summary>`);
const summary = editor.dom.select(`[data-mce-id="${uid"]`);
if (summary) {
  editor.selection.setCursorLocation(summary, 1);
  summary.removeAttribute('data-mce-id');
}
```

I think we should consider that since it would reduce the complexity here a ton since the insert code is already handling a lot of these things. At least try that apporach if it failes then we can do something else.

Reply:
Can you toggle the enabled state of the insert button based on the node changed event? This is how other features block nested inserting, I believe.

Reply:
Thanks for the tip! Disabled the button if the cursor is within the <summary> element. I've tried using `sugar` but I do not feel that it is cleaner, all the SugarElement conversions make this verbose too.